### ISSUE TITLE
Creating a `CHECK` constraint inside an `OBJECT` fails

### ISSUE THREAD

**Issue by hammerhead** [crate/crate#14341]
### CrateDB version

5.3.3 / 5.4.0-SNAPSHOT-aa3f67c (latest at the time of writing)

### Problem description

Creating a table with a `CHECK` constraint inside an `OBJECT` fails with a data type mismatch.

### Steps to Reproduce

```sql
CREATE TABLE q3 (
  a OBJECT AS (
    b INTEGER CHECK (b < 100) -- not sure if the column needs to referred to as a['b'] in the expression, but that produces the same error
  )
);
```

### Actual Result

```
SQLParseException[Cannot cast `'b'` of type `text` to type `integer`]

io.crate.exceptions.SQLParseException: Cannot cast `'b'` of type `text` to type `integer`
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:162)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:151)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:163)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:119)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:887)
	at java.base/java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2357)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:117)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:81)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.protocols.http.HttpBlobHandler.channelRead0(HttpBlobHandler.java:166)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.crate.auth.HttpAuthUpstreamHandler.handleHttpChunk(HttpAuthUpstreamHandler.java:135)
	at io.crate.auth.HttpAuthUpstreamHandler.channelRead0(HttpAuthUpstreamHandler.java:84)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:397)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: io.crate.exceptions.ConversionException: Cannot cast `'b'` of type `text` to type `integer`
	at io.crate.expression.scalar.cast.ImplicitCastFunction.normalizeSymbol(ImplicitCastFunction.java:146)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.normalizeFunction(EvaluatingNormalizer.java:238)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:228)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:123)
	at io.crate.expression.symbol.Function.accept(Function.java:197)
	at io.crate.expression.symbol.FunctionCopyVisitor.twoArgs(FunctionCopyVisitor.java:93)
	at io.crate.expression.symbol.FunctionCopyVisitor.processAndMaybeCopy(FunctionCopyVisitor.java:57)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.normalizeFunction(EvaluatingNormalizer.java:235)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:228)
	at io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:123)
	at io.crate.expression.symbol.Function.accept(Function.java:197)
	at io.crate.expression.eval.EvaluatingNormalizer.normalize(EvaluatingNormalizer.java:264)
	at io.crate.analyze.expressions.ExpressionAnalyzer.convert(ExpressionAnalyzer.java:249)
	at io.crate.analyze.CreateTableStatementAnalyzer.lambda$analyze$0(CreateTableStatementAnalyzer.java:68)
	at io.crate.sql.tree.CheckColumnConstraint.map(CheckColumnConstraint.java:87)
	at io.crate.sql.tree.ColumnDefinition.lambda$map$0(ColumnDefinition.java:160)
	at io.crate.common.collections.Lists2.map(Lists2.java:108)
	at io.crate.sql.tree.ColumnDefinition.map(ColumnDefinition.java:160)
	at io.crate.sql.tree.ObjectColumnType.lambda$map$0(ObjectColumnType.java:64)
	at io.crate.common.collections.Lists2.map(Lists2.java:108)
	at io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:64)
	at io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:32)
	at io.crate.analyze.CreateTableStatementAnalyzer.analyze(CreateTableStatementAnalyzer.java:105)
	at io.crate.analyze.Analyzer$AnalyzerDispatcher.visitCreateTable(Analyzer.java:392)
	at io.crate.analyze.Analyzer$AnalyzerDispatcher.visitCreateTable(Analyzer.java:236)
	at io.crate.sql.tree.CreateTable.accept(CreateTable.java:76)
	at io.crate.analyze.Analyzer.analyze(Analyzer.java:219)
	at io.crate.action.sql.Session.analyze(Session.java:317)
	at io.crate.action.sql.Session.parse(Session.java:307)
	at io.crate.rest.action.SqlHttpHandler.executeSimpleRequest(SqlHttpHandler.java:236)
	at io.crate.rest.action.SqlHttpHandler.handleSQLRequest(SqlHttpHandler.java:207)
	at io.crate.rest.action.SqlHttpHandler.channelRead0(SqlHttpHandler.java:116)
	... 43 more
```

### Expected Result

Successful table creation.

**Comment by matriv** [comment]
Unfortunately this exists since `4.2.0` when the `CHECK` constraints where introduced.

**Comment by BaurzhanSakhariev** [comment]
I think it was working on 4.8 and 5.0 then was one more regression probably. 
https://github.com/crate/crate/issues/12936

**Comment by matriv** [comment]
Unfortunately, this issue still exists also for ALTER TABLE:
```
@Test
public void test_check_constraint_can_be_added_to_nested_non_existing_object_sub_column() throws Exception {
    e = SQLExecutor.builder(clusterService)
        .addTable("create table t (i int, o object)")
        .build();

    analyze("alter table t add column o2 object as (b string check (b > 100))");
}
```
It works "accidentally" for string/text columns, but not for others, since the columns are casted to string literals and during normalization the casting from string to integer (in this example) fails:
```
Cannot cast `'b'` of type `text` to type `integer`
io.crate.exceptions.ConversionException: Cannot cast `'b'` of type `text` to type `integer`
	at __randomizedtesting.SeedInfo.seed([DCBA460C3C139739:7D3D997E3457109D]:0)
	at app//io.crate.expression.scalar.cast.ImplicitCastFunction.normalizeSymbol(ImplicitCastFunction.java:146)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.normalizeFunction(EvaluatingNormalizer.java:238)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:228)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:123)
	at app//io.crate.expression.symbol.Function.accept(Function.java:197)
	at app//io.crate.expression.symbol.FunctionCopyVisitor.twoArgs(FunctionCopyVisitor.java:93)
	at app//io.crate.expression.symbol.FunctionCopyVisitor.processAndMaybeCopy(FunctionCopyVisitor.java:57)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.normalizeFunction(EvaluatingNormalizer.java:235)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:228)
	at app//io.crate.expression.eval.EvaluatingNormalizer$BaseVisitor.visitFunction(EvaluatingNormalizer.java:123)
	at app//io.crate.expression.symbol.Function.accept(Function.java:197)
	at app//io.crate.expression.eval.EvaluatingNormalizer.normalize(EvaluatingNormalizer.java:264)
	at app//io.crate.analyze.expressions.ExpressionAnalyzer.convert(ExpressionAnalyzer.java:250)
	at app//io.crate.analyze.CreateTableStatementAnalyzer.lambda$analyze$0(CreateTableStatementAnalyzer.java:68)
	at app//io.crate.sql.tree.CheckColumnConstraint.map(CheckColumnConstraint.java:87)
	at app//io.crate.sql.tree.ColumnDefinition.lambda$map$0(ColumnDefinition.java:160)
	at app//io.crate.common.collections.Lists2.map(Lists2.java:108)
	at app//io.crate.sql.tree.ColumnDefinition.map(ColumnDefinition.java:160)
	at app//io.crate.sql.tree.ObjectColumnType.lambda$map$0(ObjectColumnType.java:67)
	at app//io.crate.common.collections.Lists2.map(Lists2.java:108)
	at app//io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:67)
	at app//io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:33)
	at app//io.crate.sql.tree.ColumnDefinition.map(ColumnDefinition.java:159)
	at app//io.crate.sql.tree.ObjectColumnType.lambda$map$0(ObjectColumnType.java:67)
	at app//io.crate.common.collections.Lists2.map(Lists2.java:108)
	at app//io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:67)
	at app//io.crate.sql.tree.ObjectColumnType.map(ObjectColumnType.java:33)
	at app//io.crate.analyze.CreateTableStatementAnalyzer.analyze(CreateTableStatementAnalyzer.java:105)
	at app//io.crate.analyze.Analyzer$AnalyzerDispatcher.visitCreateTable(Analyzer.java:392)
	at app//io.crate.analyze.Analyzer$AnalyzerDispatcher.visitCreateTable(Analyzer.java:236)
	at app//io.crate.sql.tree.CreateTable.accept(CreateTable.java:76)
	at app//io.crate.analyze.Analyzer.analyze(Analyzer.java:219)
	at app//io.crate.testing.SQLExecutor.analyze(SQLExecutor.java:816)
	at app//io.crate.testing.SQLExecutor.analyze(SQLExecutor.java:825)
	at app//io.crate.analyze.CreateAlterTableStatementAnalyzerTest.analyze(CreateAlterTableStatementAnalyzerTest.java:128)
	at app//io.crate.analyze.CreateAlterTableStatementAnalyzerTest.analyze(CreateAlterTableStatementAnalyzerTest.java:123)
	at app//io.crate.analyze.CreateAlterTableStatementAnalyzerTest.test_check_constraint_cannot_be_added_to_nested_object_sub_column(CreateAlterTableStatementAnalyzerTest.java:1434)
```

**Comment by matriv** [comment]
Since this is broken since the introduction of `CHECK` constraints, back at `4.2.0`, the plan is to forbid those definitions at "column level", (still can be added at table level - so there is a workaround), and open a new feature issue to support them in the future.

**Comment by BaurzhanSakhariev** [comment]
Hi @matriv, thanks for opening https://github.com/crate/crate/issues/14355 to track declaring CHECK constraint at the sub-column as a feature and providing a workaround.

https://github.com/crate/crate/issues/14355 is completed and sub-column CHECK declarations will be available in 5.5
